services/media/service/migrations/committed/000026-next-inbox-message-ingest-priority-added.sql

```diff
@@ -0,0 +1,112 @@
+--! Previous: sha1:9e58d9212f8204cc42f4e3b3afd0abe7d343bd12
+--! Hash: sha1:7f4ccc8d5137091cfb2747dd7f3dfb1bd3ac7117
+--! Message: next-inbox-message-ingest-priority-added
+
+-- Create the function to get the next batch of messages from the inbox table.
+DROP FUNCTION IF EXISTS app_hidden.next_inbox_messages(integer, integer);
+CREATE OR REPLACE FUNCTION app_hidden.next_inbox_messages(
+  max_size integer, lock_ms integer)
+    RETURNS SETOF app_hidden.inbox 
+    LANGUAGE 'plpgsql'
+
+AS $BODY$
+DECLARE 
+  loop_row app_hidden.inbox%ROWTYPE;
+  message_row app_hidden.inbox%ROWTYPE;
+  ids uuid[] := '{}';
+BEGIN
+
+  IF max_size < 1 THEN
+    RAISE EXCEPTION 'The max_size for the next messages batch must be at least one.' using errcode = 'MAXNR';
+  END IF;
+
+  -- get 1 oldest not locked priority ingest message, either StartIngest or CheckFinishIngestDocument
+  BEGIN
+    SELECT *
+      INTO message_row
+      FROM app_hidden.inbox
+      WHERE processed_at IS NULL AND abandoned_at IS NULL AND aggregate_type = 'ingest-document'
+      ORDER BY created_at
+      LIMIT 1
+      FOR NO KEY UPDATE NOWAIT; -- throw/catch error when locked
+    
+    IF FOUND AND message_row.locked_until <= NOW() THEN
+      ids := array_append(ids, message_row.id);
+    END IF;
+  EXCEPTION 
+    WHEN lock_not_available THEN
+        NULL;
+    WHEN serialization_failure THEN
+        NULL;
+    WHEN OTHERS THEN
+        RAISE;
+  END;
+
+  -- get (only) the oldest message of every segment but only return it if it is not locked
+  FOR loop_row IN
+    SELECT * FROM app_hidden.inbox m WHERE m.id in (SELECT DISTINCT ON (segment) id
```

Comment:
Can you try combining the query in line 25 with this one e.g. via union? Something like this (not tested):

```
(SELECT * FROM app_hidden.inbox m WHERE m.id in (SELECT DISTINCT ON (segment) id
  FROM app_hidden.inbox
  WHERE processed_at IS NULL AND abandoned_at IS NULL AND aggregate_type = 'ingest-document'
  ORDER BY segment, created_at) order by created_at limit 1)
  
UNION

(SELECT * FROM app_hidden.inbox m WHERE m.id in (SELECT DISTINCT ON (segment) id
  FROM app_hidden.inbox
  WHERE processed_at IS NULL AND abandoned_at IS NULL
  ORDER BY segment, created_at) order by created_at)
```

Reply:
As discussed, no need to do this. Instead, appropriate messages now have a priority segment assigned to them.